samples/core/callback/main.cpp

```diff
@@ -0,0 +1,341 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// OpenCL SDK includes
+#include <CL/SDK/CLI.hpp>
+#include <CL/SDK/Context.hpp>
+#include <CL/SDK/Image.hpp>
+#include <CL/SDK/Options.hpp>
+#include <CL/Utils/Context.hpp>
+
+// standard header includes
+#include <cassert>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <tuple> // std::make_tuple
+#include <vector>
+
+// TCLAP includes
+#include <tclap/CmdLine.h>
+
+namespace {
+
+struct ReadJob
+{
+    // Although the buffer is not used in the callback, we keep a reference to
+    // the OpenCL object to keep it alive while the read is performed.
+    cl::Buffer buffer;
+    std::vector<cl_uchar> data;
+    std::size_t side;
+};
+
+std::mutex read_mutex;
+std::map<std::intptr_t, ReadJob> read_jobs;
+std::vector<std::future<void>> image_write_tasks;
+
+template <typename T> struct DoubleBuffer
+{
+    T read, write;
+
+    void swap() { std::swap(read, write); }
+};
+
+void CL_CALLBACK read_complete_callback(cl_event, cl_int, void* user_data)
+{
+    // Acquire the job associated with the received ID and remove it from the
+    // map. These operations have to be protected, because the main thread or
+    // other callbacks might modify the map concurrently.
+    const auto job_id = reinterpret_cast<std::intptr_t>(user_data);
+    ReadJob job;
+    {
+        std::lock_guard<std::mutex> copy_lock(read_mutex);
+        const auto iter = read_jobs.find(job_id);
+        job = std::move(iter->second);
+        read_jobs.erase(iter);
+    }
+
+    // Launch an asynchronous task (on the CPU) which writes the read contents
+    // of the buffer to a PNG file. Notice, how the relevant members of the job
+    // are moved to the closure.
+    // Since no std::launch is passed, the task will either run deferred (on the
+    // main thread calling .wait()), or truly asynchronously on a different
+    // thread. The main point is that it is not running on the OpenCL runtime's
+    // thread.
+    image_write_tasks[static_cast<std::size_t>(job_id)] =
+        std::async([pixels = std::move(job.data), side = job.side, job_id] {
+            const cl::sdk::Image image{
+                static_cast<int>(side),
+                static_cast<int>(side),
+                static_cast<int>(sizeof(cl_uchar4)),
+                std::move(pixels),
+            };
+            const std::string filename =
+                "callbackcpp_out" + std::to_string(job_id) + ".png";
+            cl::sdk::write_image(filename.c_str(), image);
```

Comment:
I quite like how the C sample prints the images it wrote.  This is a good way to show progress and it hints to the user to look for images that were written.  Could a similar output be added to the C++ version also?

Note: it'd be even better to print the file names for the images that are written, for both the C and C++ versions!

Reply:
Agreed. Fixed.